add boilerplate function and use it to intercept/enrich trigger exception loops…

odain-cbd opened the pull request:
we had difficulties to identify why a notification on timeout event was not sent. we focused on the wrong trigger. another one trigger a PHP exception and lost us. it took time to find the root cause of such a ticket. 

I propose to intercept the exception and enrich it with the ticket classname and its id. 

boilerplate function can be used at other places of the code....

odain-cbd committed:
N°3265 - Log stacktrace when cron exception raised with debug enabled

odain-cbd committed:
add boilerplate function and use it to intercept/enrich trigger exception loops

accognet committed:
N°3234 - Notify on expiration not sending notifications

odain-cbd committed:
EnrichRaisedException: parameter renamed + prevent non valid objects

eespie committed:
N°3078 - Fix error on login while in maintenance mode

eespie committed:
N°3188 - Fix LIfeCycle visualization details
Fix js broken by the fix of N°309

eespie committed:
N°3162 - Remove default admin phone number (can be incompatible with validation pattern)

Commit:
:greenheart: test readability

accognet committed:
N°3261 - Configure this list : sort icon disappears when descending sort is selected

accognet committed:
N°3146 - Affichage des class user dans l'import CSV

accognet committed:
N°3149 - Change Color of Brick Search on Portail with extension Custom

accognet committed:
N°3163 - Portal Filters doesn't work

eespie committed:
N°3260 - Fix rendering of an ExternalField on a Text with XML content (format transitivity)

eespie committed:
N°3260 - Fix rendering of an ExternalField on a Text with XML content (format transitivity)

Commit:
Un-deprecate ItopExtensionsExtraRoutes::AddRoutes(...)

we had imagined it as a compatibility layer for migrating from Silex to Symfony,
but it must'nt be deprecated until we provide a new API (ideally based on a scan of yaml configuration file?)

eespie committed:
N°3285 - Fix Standard Global Search: multiple words search

eespie committed:
N°3238 - Fix multi-words search in FilterBrick and ManageBrick

accognet committed:
3234 - php 5.6 compatibility

piRGoif commented on application/utils.inc.php:
```diff
@@ -2316,4 +2316,27 @@ public static function ToCamelCase($sInput)
+		}
+
+		$sMessage = $oException->getMessage() . " (" . get_class($oCmdbAbstract) . "::" . $oCmdbAbstract->Get('id') . ")";
+		$reflectedObject = new \ReflectionClass(get_class($oException));
+		$property = $reflectedObject->getProperty('message');
+		$property->setAccessible(true);
+		$property->setValue($oException, $sMessage);
+		$property->setAccessible(false);
+		throw $oException;
```
This seems quite weird
Why not encapsulate exception as this is commonly done ? 
Having for example a TriggerException and doing something like :
`return new TriggerException($oTrigger, $oException);`

odain-cbd replied:
i agree with you. that's what i initially intended to do (encapsulating the original exception). 

but i'd prefer to keep the original exception to avoid to generate any regression in iTop core. somewhere we may rely on a specific kind of exception to do specific additional treatment...

last, we agreed with quality team to keep that PR in next 2.7.2. only modifying exception message attribute is safer.

odain-cbd replied:
as discussed with Eric we will keep this change as is for the below reasons.

piRGoif replied:
Depends on what catch we are talking about O:)
Triggers are executed by iTop Core, so we should identify which catch are triggered. I bet they are `catch(Exception $e)` ? If so, throwing a typed expression won't affect iTop, and it will be far better for future uses (a specific catch when launching triggers).

We might need to discuss this directly ?

Hipska replied:
And what about making `Trigger::DoActivate` to not throw exceptions? But return a true/false on success? 

It seems it is always used inside a loop.

piRGoif replied:
As trigger are written in modules, they can throw any exception, we don't have any control on this :/
This PR aims to add some traces that helps when doing support, but without changing any behavior so that we can commit to a support branch (warning @odain-cbd you're based on develop by the way !!). We will certainly improves trigger exception catching in the next release (develop branch), but that will be done apart.

piRGoif replied:
Here is the catch I found !

Using the [Crud stack documentation](https://www.itophub.io/wiki/page?id=latest%3Acustomization%3Asequence_crud), we can see that triggers are launched by DBObject::DBInsertNoReload and DBObject::DBUpdate.

Here are the call stack we doing an insert or an update from the console UI :

```
UI.php operation=apply_new
  catch(CoreException $e)
  catch(Exception $e)
    catch (CoreCannotSaveObjectException $e)
      cmdbAbstractObject::DBInsert
        cmdbAbstractObject::DBInsertNoReload
          DBObject::DBInsertNoReload

UI.php operation=apply_modify
  catch(CoreException $e)
  catch(Exception $e)
    catch (CoreCannotSaveObjectException $e)
    catch (DeleteException $e)
      cmdbAbstractObject::DBUpdate
        CMDBObject::DBUpdate
          DBObject::DBUpdate
```

The portal does it using this stack :

```
\Combodo\iTop\Portal\Form\ObjectFormManager::OnSubmit
  \DBObject::DBWrite
    \DBObject::DBInsert // if m_bIsInDB
    \DBObject::DBUpdate // otherwise
```

There are no catch here at all.

So as a conclusion there are no place where we have very specific catch, and we can use exception encapsulation.

Hipska replied:
Looks much better now!

odain-cbd replied:
thanks. yes code seems nicer. but stack trace seems worst to me now. as we loose the root cause of the original exception raised. this is not final code.

Hipska replied:
It's not lost, it's only not represented. I think it's just a matter of calling `$this->getPrevious()->GetTrace()` inside `CoreException::getTraceAsHtml()` to be added into `$aBackTrace`.

odain-cbd replied:
thanks i'll try

odain-cbd replied:
chaining the exception is nicer. as you proposed (Hipska), I can adapt CoreException->getTraceAsHtml() to keep previous stack trace. but i confirm that stacktrace is lost in getTraceAsString(). Reflection manipulation remains better to me.

https://stackoverflow.com/questions/31562724/php-chaining-error-handling

https://stackoverflow.com/questions/34923644/does-re-throwing-an-exception-in-php-destroy-the-stack-trace

Hipska replied:
These are 2 irrelevant links, see this example to prove the stack trace is not lost: https://3v4l.org/NnPbt
It's really is only matter of correctly retrieving the trace.

odain-cbd replied:
trace is not lost when using __tostring($e) of your chained exception. on the other hand we dont have root cause when using getTrace() or getTraceAsString(). this is really annnoying as we cannot override them (final public methods).

as discussed with Pierre, we will keep your proposal (chaining the exception) and add a new method getFullStackTraceAsString().

odain-cbd committed:
add new code style for brackets

accognet committed:
N°3303 - Bug on Mass update of actions (notification)

Hipska commented on application/utils.inc.php:
```diff
@@ -2316,4 +2316,27 @@ public static function ToCamelCase($sInput)
+	public static function EnrichRaisedException($oCmdbAbstract, $oException)
+	{
+		if (is_null($oCmdbAbstract) ||
+			! is_a($oCmdbAbstract, 'cmdbAbstractObject'))
```
```suggestion
			! is_a($oCmdbAbstract, \cmdbAbstractObject::class))
```

piRGoif replied:
Thanks Hipska, this is indeed better for IDEs !
And this is compatible PHP 5.5 : https://www.php.net/manual/fr/language.oop5.basic.php#language.oop5.basic.class.class
And was already used in iTop

piRGoif committed:
:memo: N°2293 PHPDoc for DBObject::GetOriginal

piRGoif committed:
:memo: Fix CRUD wiki page URL
was linking to Combodo private wiki :/, now is the public one \o/
Many thanks @Hipska !

odain-cbd committed:
Update application/utils.inc.php

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

odain-cbd committed:
encapsulation proposal

Hipska commented on core/coreexception.class.inc.php:
```diff
@@ -28,7 +28,7 @@
 class CoreException extends Exception
 {
-	public function __construct($sIssue, $aContextData = null, $sImpact = '')
+	public function __construct($sIssue, $aContextData = null, $sImpact = '', Throwable $previous = null)
```
```suggestion
	public function __construct($sIssue, $aContextData = null, $sImpact = '', Throwable $oPrevious = null)
```

Hipska replied:
Same here @odain-cbd @piRGoif . Oh, and it has type-hinting ;)

piRGoif replied:
And also there is type hinting... And Throwable seems to have been added in PHP 7 only (https://www.php.net/manual/fr/class.throwable.php)

odain-cbd replied:
type hinting removed

Hipska commented on core/coreexception.class.inc.php:
```diff
@@ -66,7 +66,7 @@ public function __construct($sIssue, $aContextData = null, $sImpact = '')
 			$sMessage .= implode(', ', $aContextItems);
 		}
-		parent::__construct($sMessage, 0);
+		parent::__construct($sMessage, 0, $previous);
```
```suggestion
		parent::__construct($sMessage, 0, $oPrevious);
```

odain-cbd replied:
yes i'll change

Hipska replied:
Although this conversation was marked as resolved, it is still not resolved.

piRGoif replied:
Conversation reopened, thanks for your notice ! I'll tell @odain-cbd

odain-cbd replied:
code cleanup done

Hipska commented on application/utils.inc.php:
```diff
@@ -2316,4 +2316,24 @@ public static function ToCamelCase($sInput)
+	 *
+	 * @throws \Exception
+	 */
+	public static function EnrichRaisedException($oCmdbAbstract, $oException)
+	{
+		if (is_null($oCmdbAbstract) ||
+			! is_a($oCmdbAbstract, \cmdbAbstractObject::class))
+		{
+			throw $oException;
+		}
```
Why not like this?
```suggestion
	public static function EnrichRaisedException(\cmdbAbstractObject $oCmdbAbstract, Throwable $oException){
		if (is_null($oCmdbAbstract)){
			throw $oException;
		}
```

odain-cbd replied:
Type hinting is a good idea. but it is not authorized until latest develop branch. ie 2.8.

as we would like to benefit from this work in 2.7.x (long term support) I will not apply your proposal. even if it is better to me.

Hipska replied:
So, starting from which version is type hinting authorised? I think there are lots of other places that already do type hinting..

piRGoif replied:
We decided to add type hinting only starting from 2.8.0 (develop branch), and only for new methods : a very conservative approach, but we wouldn't like to add bugs because of this...

piRGoif replied:
Note that the [coding standard wiki page](https://www.itophub.io/wiki/page?id=latest%3Acustomization%3Acoding_standards) wasn't updated yet, as this decision was made only few weeks ago, and we want to test drive this internally first...

Hipska replied:
Well, correct me if I'm wrong, but isn't this a new method?

piRGoif replied:
There are 2 conditions to add type hinting :
* code starting on the 2.8.0
* new method

We are adding a new method but on a previous version.

For now this PR is based on develop but this was a mistake and will be fixed very soon : it will be merged to the support/2.7 branch.

Hipska replied:
Oh, okay. That was not clear. 👍

Hipska commented on application/utils.inc.php:
```diff
@@ -2316,4 +2316,24 @@ public static function ToCamelCase($sInput)
+			throw $oException;
+		}
+
+		$sMessage = $oException->getMessage() . " (" . get_class($oCmdbAbstract) . "::" . $oCmdbAbstract->Get('id') . ")";
```
Why not use predefined `__toString` method?

> HTML String representation of the object
> 
> Only a few meaningful information will be returned.
> This representation is for debugging purposes, and is subject to change.
> The returned string is raw HTML

```suggestion
		$sMessage = $oCmdbAbstract . " " . $oException->getMessage();
```

odain-cbd replied:
good idea for code refactoring. but __toString() currently returns html content. isn't it weard in logs?

Hipska replied:
True, but that is a weirdness of the `__toString` method.

Note that most of these messages do get displayed on the UI (Hence the existence of `CoreException::getTraceAsHtml()`)

odain-cbd replied:
i will used __tostring() as recommanded.

Hipska replied:
I didn't recommend it. I was just asking if this wouldn't be the better way.

piRGoif committed:
Update .editorconfig

piRGoif committed:
:art: Fix string delimiter
Thanks @jbostoen for pointing this out (see discussion in 0773660ef26164a8ab24a151817bb148fe368f08)

piRGoif committed:
:wrench: Fix editorconfig for JS files

piRGoif committed:
:art: Fix invalid formatting introduced in 827b4b5bbe9613104fd9390db0756ab2611a1595

piRGoif committed:
:wrench: Remove versionned PHPStorm files
Those files were duplicates of the .editorconfig file

.editorconfig was pushed with ecd8f40c0f1c692f94e2f79d11be842e2d41eec4 for 2.7.0 and should be the only source of the IDE / editor configuration

This generic format is implemented in PHPStorm since v2019.2 (see https://blog.jetbrains.com/idea/2019/06/managing-code-style-on-a-directory-level-with-editorconfig/)
To check if your IDE / editor supports it, check https://editorconfig.org/

eespie committed:
N°3311 - Stop capturing output before sending backup file (avoid memory problem)

eespie committed:
N°3317 - Add http headers

eespie committed:
N°3318 - don't display error details

eespie committed:
N°3283 - Spelling mistake

eespie committed:
N°3309 - Spelling mistake

odain-cbd committed:
use cmdAbstract tostirng() for enrichment + add CoreException::getFullStackTraceAsString to keep all previous from a chained CoreException

Hipska commented on pages/UI.php:
```diff
@@ -1997,7 +1997,7 @@ function DisplayNavigatorGroupTab($oP)
 				$oLog->Set('userinfo', '');
 				$oLog->Set('issue', $e->GetIssue());
 				$oLog->Set('impact', 'Page could not be displayed');
-				$oLog->Set('callstack', $e->getTrace());
+				$oLog->Set('callstack', $e->getFullStackTraceAsString());
```
Previous method returned an array and now a string is returned. This is a huge difference.

piRGoif replied:
You're right ! But the value is written to a string field anyway.

Hipska replied:
Correct, but the structure of it will be completely different.

piRGoif replied:
Right again, but this is debug traces and we though changing format won't matter.

Hipska commented on pages/UI.php:
```diff
@@ -2007,7 +2007,7 @@ function DisplayNavigatorGroupTab($oP)
 		}
 
-		IssueLog::Error('UI.php operation='.$operation.', error='.$e->getMessage()."\n".$e->getTraceAsString());
+		IssueLog::Error('UI.php operation='.$operation.', error='.$e->getMessage()."\n".$e->getFullStackTraceAsString());
```
Aren't you over-complicating thing?
```suggestion
		IssueLog::Error('UI.php operation='.$operation.', error='.$e->getMessage()."\n".$e);
```
Or this as the toString does give the message as well:
```suggestion
		IssueLog::Error('UI.php operation='.$operation.', error='.$e);
```

piRGoif replied:
Adding a new method allows us to have a clear PHPDoc explaining why we're doing this. And we're also not totally confident that Exception::toString will suit our needs in the future.

Hipska replied:
Okay, can be solved for me.

piRGoif commented on core/coreexception.class.inc.php:
```diff
@@ -28,7 +28,7 @@
 
 class CoreException extends Exception
 {
-	public function __construct($sIssue, $aContextData = null, $sImpact = '')
```
There are some CoreException children that redefines the constructor, but we chose not to update them yet, as this modification aims to be merged to support/2.7 branch.
The work will continue later on, actually handling exceptions throwned by non iTop core code (modules, datamodel methods overrides, ...) is a vast subject O:)

piRGoif commented on core/coreexception.class.inc.php:
```diff
@@ -81,6 +81,16 @@ public function getHtmlDesc($sHighlightHtmlBegin = '<b>', $sHighlightHtmlEnd = '
 	}
 
+	/**
+	 * getTraceAsString() cannot be overrided and it is limited as only current exception stack is returned.
+	 * we need stack of all previous exceptions
```
Having a PSR-3 compatible logger would allow to get rid of this... In such logger the exception is passed directly to the log method, whereas iTop logger only accept a string so logger consumer needs to call itself the getTrace() method.

We might need to add this to LogAPI, will discuss this with the team...

odain-cbd committed:
previous renaming + remove Throwable type hinting

odain-cbd committed:
renaming to

odain-cbd commented:
PR cherry-picked and commited in support/2.7 (4e0eed6e1).

Thanks for the active collaboration of all reviewers. we went deeper than I would have imagined!
